File: apps/ios/Voiid/Voiid/Games/Ludo/LudoTurnBorder.swift
//
//  LudoTurnBorder.swift
//  Voiid
//
//  Turn-border sweep (§12). The perimeter is a rounded rectangle inset by half the stroke;
//  anchors are normalized clockwise fractions: red 0.00 bottom-left, green 0.25 top-left,
//  yellow 0.50 top-right, blue 0.75 bottom-right.
//
//  CONSTRUCTION (§12.2): the OLD hue stays as the base full stroke; a new-hue OVERLAY travels
//  from the outgoing anchor to the incoming anchor using Path trim with WRAPPED segments when
//  the phase crosses path length. Rounded cap while traveling, butt at completion. Never two
//  cross-faded borders, never a gradient disguise.
//

import SwiftUI

enum LudoTurnBorder {

    /// Phase from seat A to seat B clockwise in [0,1): adjacent 0.25, duel opposite 0.5, wrap 0.75.
    static func sweepPhase(from: Int, to: Int) -> CGFloat {
        let delta = ((to - from) % 4 + 4) % 4
        return CGFloat(delta) / 4
    }

    /// The perimeter path: rounded rect inset by half its stroke so the stroke stays inside.
    static func perimeterPath(side: CGFloat, cornerRadius: CGFloat, stroke: CGFloat) -> Path {
        let inset = stroke / 2
        let rect = CGRect(x: inset, y: inset,
                          width: side - stroke, height: side - stroke)
        return Path(roundedRect: rect, cornerRadius: min(cornerRadius, side / 2))
    }

    /// One frame of the border.
    ///
    /// - Parameters:
    ///   - baseColor: steady full border under everything (old hue, or podBorder waiting)
    ///   - overlayColor: traveling new hue, or nil when resting
    ///   - phaseStart: normalized start anchor of travel
    ///   - progress: 0...1 along the FULL perimeter from phaseStart (eased by the caller)
    static func draw(
        _ ctx: inout GraphicsContext,
        path: Path,
        stroke: CGFloat,
        baseColor: Color,
        overlayColor: Color?,
        phaseStart: CGFloat,
        progress: CGFloat,
    ) {
        ctx.stroke(path, with: .color(baseColor),
                   style: StrokeStyle(lineWidth: stroke, lineCap: .butt))

        guard let overlay = overlayColor, progress > 0 else { return }
        if progress >= 1 {
            // Completion REPLACES the base hue entirely (§12.2).
            ctx.stroke(path, with: .color(overlay),
                       style: StrokeStyle(lineWidth: stroke, lineCap: .butt))
            return
        }

        // Traveling segment may wrap past path length; draw up to two sub-trims. Rounded
        // leading cap during travel reads as motion; completion switches to butt above.
        let tail = phaseStart.truncatingRemainder(dividingBy: 1)
        let head = (tail + progress).truncatingRemainder(dividingBy: 1)
        if tail < head {
            ctx.stroke(trimmed(path, from: tail, to: head), with: .color(overlay),
                       style: StrokeStyle(lineWidth: stroke, lineCap: .round))
        } else {
            ctx.stroke(trimmed(path, from: tail, to: 1), with: .color(overlay),
                       style: StrokeStyle(lineWidth: stroke, lineCap: .round))
            ctx.stroke(trimmed(path, from: 0, to: head), with: .color(overlay),
                       style: StrokeStyle(lineWidth: stroke, lineCap: .round))
        }
    }

    private static func trimmed(_ path: Path, from: CGFloat, to: CGFloat) -> Path {
        guard to > from else { return Path() }
        return path.trimmedPath(from: from, to: to)
    }
}

/// The pod timer ring (§13). Server-timed: renders `deadlineAt − estimatedServerNow`, clamped
/// 0…30 s, starting at 100% at opensAt and depleting clockwise from 12 o'clock. Player hue
/// until 5 s, timerWarning at 5 s, timerCritical at 2 s — HOW LONG only; never WHO.
enum LudoTimerRing {

    struct RingState {
        let fractionRemaining: Double
        let secondsRemaining: Int
        var colorOverride: Color?   // warning/critical; nil → player hue
    }

    /// Pure math over the SMOOTHED server clock (never a raw device counter).
    static func state(opensAt: Double?, deadlineAt: Double?, estimatedNowMs: Double) -> RingState? {
        guard let opensAt, let deadlineAt else { return nil }
        let total = max(deadlineAt - opensAt, 1)
        let remaining = min(max(deadlineAt - estimatedNowMs, 0), LudoRules.turnWindowMs)
        let seconds = Int(remaining / 1000)
        let colorOverride: Color? = {
            if remaining <= 2_000 { return Color(hex: 0xEF7A6B) }   // resolved by theme below
            if remaining <= 5_000 { return Color(hex: 0xB07818) }
            return nil
        }()
        _ = colorOverride
        return RingState(
            fractionRemaining: remaining / total,
            secondsRemaining: seconds,
            colorOverride: nil) // theme-aware override applied by LudoPlayerPod
    }

    static func draw(
        _ ctx: inout GraphicsContext,
        diameter: CGFloat,
        stroke: CGFloat,
        track: Color,
        arc: Color,
        fraction: Double,
    ) {
        let rect = CGRect(x: 0, y: 0, width: diameter, height: diameter).insetBy(dx: stroke / 2, dy: stroke / 2)
        ctx.stroke(Path(ellipseIn: rect), with: .color(track), lineWidth: stroke)
        let f = min(max(fraction, 0), 1)
        guard f > 0 else { return }
        var arcPath = Path()
        arcPath.addArc(center: CGPoint(x: diameter / 2, y: diameter / 2),
                       radius: diameter / 2 - stroke / 2,
                       startAngle: .degrees(-90),
                       endAngle: .degrees(-90 - 360 * f),
                       clockwise: true)
        ctx.stroke(arcPath, with: .color(arc),
                   style: StrokeStyle(lineWidth: stroke, lineCap: .round))
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/games/ludo/LudoTurnBorder.kt
package com.voiid.app.main.games.ludo

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.asAndroidPath
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke

/**
 * Turn-border sweep (§12). The perimeter is a rounded rectangle inset by half the stroke;
 * anchors are normalized clockwise fractions: red 0.00 bottom-left, green 0.25 top-left,
 * yellow 0.50 top-right, blue 0.75 bottom-right.
 *
 * CONSTRUCTION (§12.2): the OLD hue stays as the base full stroke; a new-hue OVERLAY travels
 * from the outgoing anchor to the incoming anchor using path trim with WRAPPED segments when
 * the phase crosses path length. Rounded leading cap during travel, butt cap at completion.
 * Never two cross-faded borders, never a gradient disguise.
 */
object LudoTurnBorder {

    /** Phase from seat A to seat B clockwise, in [0,1): adjacent 0.25, duel opposite 0.5, wrap 0.75. */
    fun sweepPhase(fromSeat: Int, toSeat: Int): Float {
        val delta = ((toSeat - fromSeat) % 4 + 4) % 4
        return delta / 4f
    }

    /**
     * Draw one frame of the border.
     *
     * @param baseColor steady full border under everything (old hue, or podBorder waiting)
     * @param overlayColor traveling new hue, or null when resting
     * @param phaseStart normalized start anchor of travel
     * @param progress 0..1 along the FULL perimeter from phaseStart (eased by the caller)
     */
    fun DrawScope.drawBorder(
        rectPath: Path,
        strokePx: Float,
        baseColor: Color,
        overlayColor: Color?,
        phaseStart: Float,
        progress: Float,
    ) {
        // Base full stroke — butt caps, continuous.
        drawPath(rectPath, baseColor, style = Stroke(strokePx))

        val overlay = overlayColor ?: return
        if (progress <= 0f) return
        if (progress >= 1f) {
            // Completion replaces the base hue entirely (§12.2).
            drawPath(rectPath, overlay, style = Stroke(strokePx))
            return
        }
        // Traveling segment may wrap past path length; draw up to two sub-strokes.
        val total = 1f
        val head = (phaseStart + progress * total) % 1f
        val tail = phaseStart % 1f
        if (tail < head) {
            trimStroke(rectPath, overlay, strokePx, tail, head, rounded = true)
        } else {
            trimStroke(rectPath, overlay, strokePx, tail, 1f, rounded = true)
            trimStroke(rectPath, overlay, strokePx, 0f, head, rounded = true)
        }
    }

    /**
     * Android has no native path-trim; PathMeasure.getSegment provides exact wrapped
     * segments (§18.2). The segment is written into the backing store of a fresh compose
     * Path via its own asAndroidPath().
     */
    /**
     * One arc of [length] (0..1 of the perimeter) starting at [phaseStart], wrapping if it runs
     * past the end of the path. Used by the turn clock, which shortens this arc from the active
     * seat's own anchor as their window runs down.
     */
    fun DrawScope.drawArc(
        path: Path,
        strokePx: Float,
        color: Color,
        phaseStart: Float,
        length: Float,
    ) {
        if (length >= 1f) {
            drawPath(path, color, style = Stroke(strokePx))
            return
        }
        val tail = phaseStart % 1f
        val head = tail + length
        if (head <= 1f) {
            trimStroke(path, color, strokePx, tail, head, rounded = false)
        } else {
            trimStroke(path, color, strokePx, tail, 1f, rounded = false)
            trimStroke(path, color, strokePx, 0f, head - 1f, rounded = false)
        }
    }

    private fun DrawScope.trimStroke(
        path: Path,
        color: Color,
        strokePx: Float,
        fromFrac: Float,
        toFrac: Float,
        rounded: Boolean,
    ) {
        if (toFrac <= fromFrac) return
        val pm = android.graphics.PathMeasure()
        pm.setPath(path.asAndroidPath(), false)
        val len = pm.length
        if (len <= 0f) return
        val start = len * fromFrac
        val stop = len * toFrac
        if (stop - start <= 0.5f) return
        val out = Path()
        if (!pm.getSegment(start, stop, out.asAndroidPath(), true)) return
        drawPath(
            out,
            color,
            style = Stroke(
                width = strokePx,
                cap = if (rounded) StrokeCap.Round else StrokeCap.Butt,
            ),
        )
    }
}
